### ISSUE TITLE
Infinite Loop in analyzeInstruction

### ISSUE THREAD

**Issue by SteveR113322** [vincent-zurczak/Xml-Region-Analyzer#3]
Infinite Loop in analyzeInstruction if XML has "<?" but no "?>"

this works for me:

boolean analyzeInstruction( String xml, List<XmlRegion> positions ) {

		boolean result = false;
		int newPos = this.offset;
		if( newPos < xml.length()
				&& xml.charAt( newPos ) == '<'
				&& ++ newPos < xml.length()
				&& xml.charAt( newPos ) == '?' ) {

			while( newPos > -1 && ++ newPos < xml.length() 
					&& xml.charAt( newPos ) != '>' )
				newPos = xml.indexOf( '?', newPos );

			if( newPos > -1 && xml.charAt( newPos ) == '>' ) {
				positions.add( new XmlRegion( XmlRegionType.INSTRUCTION, this.offset, newPos + 1 ));
				this.offset = newPos + 1;
				result = true;
			}
		}

		return result;
	}

And thanks for sharing!!!

**Comment by SteveR113322** [comment]
Hi,

yes of course. I created a unit test for the Problem.

I can‘t commit it due to permissions but i dont think that would be the correct way.

Send the pull request to me and i will provide the fix.

Regards,

Stefan


Von: Vincent Zurczak [mailto:notifications@github.com]
Gesendet: Sonntag, 20. Mai 2018 19:22
An: vincent-zurczak/Xml-Region-Analyzer <Xml-Region-Analyzer@noreply.github.com>
Cc: Retzlaff, Stefan, A-SCM-DM-IT-E <stefan.retzlaff@arvato.com>; Author <author@noreply.github.com>
Betreff: Re: [vincent-zurczak/Xml-Region-Analyzer] Infinite Loop in analyzeInstruction (#3)


Hi. :)

Sorry for the delay. Would it be possible for you to add a unit test and provide your solution through a pull request? Thanks for the report.

**Comment by vincent-zurczak** [comment]
Closed by #4

**Comment by vincent-zurczak** [comment]
Hi there,

I am currently converting the project to Maven.
And I reviewed your code once again. I did not see it at first glance, but the sample you used in your unit test is an invalid XML.

Indeed, `<?xml version="1.0" encoding="UTF-8">` should end with `?>`, not just `>`.
Are you sure about your use case? There were already unit tests for such cases (empty string, white spaces only, XML instruction, XML instruction with surrounding white spaces).
